### PR TITLE
Desktop: Batch delete for Notebooks added

### LINKED ISSUE #2703: Desktop: Deleting Notebook Slower Than Selecting All Notes and Hitting Delete
## Environment

Joplin version: all versions past v1.0.179 (don't know about previous versions) 
Platform: Linux
OS specifics: Artix Linux 64 bit

## Steps to reproduce

1. Create a new notebook (any name will do)
2. Create 4 different notes
3. Duplicate these notes as many times as you can (i did over 1000 in my test)
4. Select 1st note 
5. Hold Shift and select last note
6. Hit Delete key or Click Delete on Right 
7. Click Yes 
8. Repeat Steps 1-3 
9. Select Notebook 
10. Repeat Step 6

## Describe what you expected to happen
Both should delete notes at the same speed but Notebook delete is drastically slower. 

## Logfile

Not home but can produce a log when i arrive.

### PR DESCRIPTION
**Pull request by RedDocMD**
Fixes #2703

### CODE REVIEW

**Comment by RedDocMD** [comment]
@tessus, is this okay?

**Comment by tessus** [comment]
Yes, looks good.

**Comment by RedDocMD** [comment]
@laurent22, can this be merged?

**Comment by mic704b** [comment]
@RedDocMD please do not send mentions for merge attention, your PR will not be lost, it will be processed in due course.

**Comment by mic704b** [comment]
@RedDocMD

Are there unit tests for `Note.batchDelete()`? I know you did not create this method, however as your change relies on it, it is good practice to add them if they do not exist.

Have you reviewed the unit tests for `Folder.delete()`.  Are there additional cases that we should add?

**Comment by RedDocMD** [comment]
@mic704b , there doesn't seem to be a unit test for `Note.batchDelete()` in 'CliClient/tests/model_Note.js`. I will add one. Also the unit test for `Folder.delete()` doesn't test for multiple notes across several sub-notebooks. So I will add that as well.

**Comment by RedDocMD** [comment]
@mic704b, I have added the tests I had mentioned above

**Comment by RedDocMD** [comment]
@PackElend, can you please add the GSoC tag for this PR?

**Comment by laurent22** [comment]
Code looks good. @mic704b feel free to merge if the tests are good to go.

File: CliClient/tests/models_Note.js [hunk 1]
```diff
@@ -129,4 +135,18 @@ describe('models_Note', function() {
 		expect(duplicatedNote.user_updated_time !== note1.user_updated_time).toBe(true);
 	}));
 
+	it('should delete a set of notebooks', asyncTest(async () => {
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: ~~notebooks~~ notes

File: CliClient/tests/models_Note.js [hunk 2]
```diff
@@ -129,4 +135,18 @@ describe('models_Note', function() {
...
+		let noOfNotes = 20;
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note1${i}`, parent_id: folder1.id });
+		}
+
+		let noteIds = await Folder.noteIds(folder1.id);
+		await Note.batchDelete(noteIds);
+
+		const all = await allItems();
+		expect(all.length).toBe(1);
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: After this, please check the item is actually the `folder1` by checking the id

File: CliClient/tests/models_Note.js [hunk 3]
```diff
@@ -129,4 +135,18 @@ describe('models_Note', function() {
...
+	}));
+
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: Please add a couple more test cases for `batchDelete`:
1. Delete notes like your test case, but not all of them.  Check that the notes that shouldn't be deleted aren't.
2. Check when noteIds is an empty list ie `batchDelete([])`.  No notes should be deleted.

**Comment by RedDocMD** [comment]
@mic704b, I have made the changes as you had instructed

File: CliClient/tests/models_Note.js [hunk 4]
```diff
@@ -129,4 +134,98 @@ describe('models_Note', function() {
...
+		let f1 = await Folder.save({ title: 'folder1' });
+		let f2 = await Folder.save({ title: 'folder2', parent_id: f1.id });
+
+		let noOfNotes = 20;
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note1${i}`, parent_id: f1.id });
+		}
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note2${i}`, parent_id: f2.id });
+		}
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: Instead use `createNTestNotes` from `test-utils.js`  (also applies to next test case)

File: CliClient/tests/models_Note.js [hunk 5]
```diff
@@ -129,4 +134,98 @@ describe('models_Note', function() {
...
+
+		const allBeforeDelete = await allItems();
+
+		let notesInFolder1IDs = await Folder.noteIds(f1.id);
+		let notesInFolder2IDs = await Folder.noteIds(f2.id);
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: Instead use `createNTestNotes` from `test-utils.js`

File: CliClient/tests/models_Note.js [hunk 6]
```diff
@@ -129,4 +134,98 @@ describe('models_Note', function() {
...
+		}
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note2${i}`, parent_id: f2.id });
+		}
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note3${i}`, parent_id: f3.id });
+		}
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note4${i}`, parent_id: f4.id });
+		}
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: Please use `createNTestNotes` for these

File: CliClient/tests/models_Note.js [hunk 7]
```diff
@@ -129,4 +134,98 @@ describe('models_Note', function() {
...
+		const beforeDelete = await allItems();
+		await Note.batchDelete([]);
+		const afterDelete = await allItems();
+
+		expect(beforeDelete.length).toBe(afterDelete.length);
+		let count = 0;
+		for (let i = 0; i < beforeDelete.length; i++) {
+			if (beforeDelete[i].id == afterDelete[i].id) ++count;
+		}
+		expect(beforeDelete.length).toBe(count);
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: Good

but you can reduce all of this to:
```
expect(sortedIds(afterDelete)).toEqual(sortedIds(beforeDelete));
```

File: CliClient/tests/models_Note.js [hunk 8]
```diff
@@ -129,4 +134,98 @@ describe('models_Note', function() {
...
+		for (let i = 0; i < noOfNotes; i++) {
+			await Note.save({ title: `note2${i}`, parent_id: f2.id });
+		}
+
+		const allBeforeDelete = await allItems();
+
+		let notesInFolder1IDs = await Folder.noteIds(f1.id);
+		let notesInFolder2IDs = await Folder.noteIds(f2.id);
+
+		const selectRandomly = (source, count) => {
```
Comment (mic704b) [inline CliClient/tests/models_Note.js]: Nice.  However lets not have a random element in the tests. Please remove and simplify this case a bit.  If you want to cover more combinations, make extra test cases explicitly.  I'd be happy with just one though, to start with.

Nice work though, you seem to know your javascript!

**Comment by RedDocMD** [comment]
@mic704b, I have made the changes you said and resolved the conflict

**Comment by mic704b** [comment]
Thanks @RedDocMD changes look good.  

You've just got a few linter problems to fix.

**Comment by RedDocMD** [comment]
@mic704b, I think I have fixed the linter issues

**Comment by mic704b** [comment]
Thanks @RedDocMD, all good.

**Pull request merged**